src/main/cpp/src/parse_uri.cu

```diff
@@ -495,7 +495,7 @@ bool __device__ validate_fragment(string_view fragment)
 __device__ std::pair<string_view, bool> find_query_part(string_view haystack, string_view needle)
 {
   auto const n_bytes     = needle.size_bytes();
-  auto const find_length = haystack.size_bytes() - n_bytes + 1;
+  auto const find_length = haystack.size_bytes() - n_bytes;
 
   auto h           = haystack.data();
   auto const end_h = haystack.data() + find_length;
```

Comment:
Usually the `end_h` should refer to the end of bytes that `h` points to.
If this assumption is true, then below code will never be out of bound.
```
while (h < end_h && *h != '&')
```
Please refactor the code, maybe like:
```cpp
__device__ std::pair<string_view, bool> find_query_part2(string_view haystack, string_view needle)
{
  auto const n_bytes = needle.size_bytes();

  auto p           = haystack.data();
  auto const end_p = haystack.data() + haystack.size_bytes();
  auto n_p         = needle.data();
  auto end_n       = needle.data() + n_bytes;

  bool find     = false;
  auto find_p   = p;
  int match_len = 0;

  while (p < end_p) {
    bool match      = false;
    size_type tried = 0;
    for (size_type jdx = 0; jdx < n_bytes && (p + jdx) < end_p && (n_p + jdx) < end_n; ++jdx) {
      if (p[jdx] != n_p[jdx]) { break; }
      tried += 1;
    }
    if (tried == n_bytes && p + n_bytes < end_p && p[n_bytes] == '=') {
      // match `needle=` in haystack
      p += n_bytes + 1;
      find       = true;
      find_p     = p;
      auto start = p;
      while (p < end_p && *p != '&') {
        ++match_len;
        ++p;
      }
      break;
    } else {
      while (p < end_p && *p != '&') {
        ++p;
      }
    }
  }

  if (find) {
    return {{find_p, match_len}, true};
  } else {
    return {{}, false};
  }
}
```

Reply:
Thanks! Refactored code, hope it looks clearer.